contracts/Datastore.sol

```diff
@@ -329,6 +341,147 @@ contract Datastore {
      * @param _entity Entity address     
      */
     function hasWriteAccess(uint _fileId, address _entity) public view returns (bool) {
-        return isOwner(_fileId, _entity) || files[_fileId].permissions[_entity].write;
+        if(isOwner(_fileId, _entity) || files[_fileId].permissions[_entity].write)
+            return true;
+        
+        for (uint i = 0; i < files[_fileId].groupIds.length; i++) {
+            if (files[_fileId].groupPermissions[files[_fileId].groupIds[i]].write) {
+                if (groups[files[_fileId].groupIds[i]].entitiesWithIndex[_entity] != 0)
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @notice Add a group to the datastore
+     * @param _groupName Name of the group
+     */
+    function createGroup(string _groupName) external {
+        uint id = groupList.length + 1;
+        require(!groups[id].exists);
+        groups[id].groupName = _groupName;
+        groups[id].exists = true;
+        groupList.push(id);
+    }
+
+    /**
+     * @notice Delete a group from the datastore
+     * @param _groupId Id of the group to delete
+     */
+    function deleteGroup(uint _groupId) external {
+        require(groups[_groupId].exists);
+        delete groups[_groupId];
+        delete groupList[_groupId - 1];
+    }
+
+    /**
+     * @notice Rename a group
+     * @param _groupId Id of the group to rename
+     * @param _newGroupName New name for the group
+     */
+    function renameGroup(uint _groupId, string _newGroupName) external  {
+        require(groups[_groupId].exists);
+        groups[_groupId].groupName = _newGroupName;
+    }
+
+    /**
+     * @notice Get a list of all the groups Id's
+     */
+    function getGroups() external view returns(uint[]){
+        return groupList;
+    }
+
+    /**
+     * @notice Get a specific group
+     * @param _groupId Id of the group to return
+     */
+    function getGroup(uint _groupId) public view returns(address[] _entities, string _groupName) {
+        require(groups[_groupId].exists);
+        _entities = groups[_groupId].entities;
+        _groupName = groups[_groupId].groupName;
+    }
+
+    /**
+     * @notice Get an entity inside a specific group
+     * @param _groupId Id of the group to retrieve the entity from
+     * @param _entityIndex Index of the entity to retrieve from the group
+     */
+    function getGroupEntity(uint _groupId, uint _entityIndex) public view returns(address) {
+        require(groups[_groupId].exists);
+        if(groups[_groupId].entities[_entityIndex] != 0)
+            return groups[_groupId].entities[_entityIndex];
+    }
+
+    /**
+     * @notice Get the number of entities in a group
+     * @param _groupId Id of the group to get the count from
+     */
+    function getGroupCount(uint _groupId) public view returns(uint) {
```

Comment:
I'm not sure the name is appropriate. At first glance, I thought this function returned the number of groups available in the app. Maybe `getGroupEntityCount` ?

Reply:
Totally agree. Will change.